### ISSUE TITLE
Running on Linux

### ISSUE THREAD

**Issue by loikki** [cagnulein/qdomyos-zwift#1121]
Hi,

Thanks a lot for your software. I have been trying to use it and was successful on android but unfortunately not so much on linux. I was able to compile and run the software, but my bluetooth connection (to my Proform CBC) is not working. Is it expected?

My bluetooth is not even stable when I try to pair with the bike without your software (but is fine with other bluetooth equipment). I am not sure if it will help, but here are my bluetooth service's logs (I tried to connect / disconnect multiple times):

```
Jan  4 22:52:07 kratos2 bluetoothd[81989]: profiles/gap/gas.c:read_device_name_cb() Reading device name failed with ATT errror: 14
Jan  4 22:52:07 kratos2 bluetoothd[81989]: profiles/gap/gas.c:read_appearance_cb() Reading appearance failed with ATT error: 14
Jan  4 22:52:07 kratos2 bluetoothd[81989]: src/device.c:att_disconnected_cb() 
Jan  4 22:52:07 kratos2 bluetoothd[81989]: src/device.c:att_disconnected_cb() Software caused connection abort (103)
Jan  4 22:52:07 kratos2 bluetoothd[81989]: src/service.c:change_state() 0x55feefa04310: device F2:1E:0F:33:13:1D profile gap-profile state changed: connected -> disconnecting (0)
Jan  4 22:52:07 kratos2 bluetoothd[81989]: src/service.c:change_state() 0x55feefa04310: device F2:1E:0F:33:13:1D profile gap-profile state changed: disconnecting -> disconnected (0)
Jan  4 22:52:07 kratos2 bluetoothd[81989]: src/gatt-client.c:btd_gatt_client_disconnected() Device disconnected. Cleaning up.
Jan  4 22:52:07 kratos2 bluetoothd[81989]: src/device.c:att_disconnected_cb() Automatic connection disabled
Jan  4 22:52:07 kratos2 bluetoothd[81989]: src/gatt-database.c:btd_gatt_database_att_disconnected() 
Jan  4 22:52:07 kratos2 bluetoothd[81989]: attrib/gattrib.c:g_attrib_unref() 0x55feef9ec3a0: g_attrib_unref=0 
Jan  4 22:52:15 kratos2 systemd[1]: systemd-hostnamed.service: Deactivated successfully.
Jan  4 22:52:17 kratos2 bluetoothd[81989]: src/device.c:device_connect_le() Connection attempt to: F2:1E:0F:33:13:1D
Jan  4 22:52:17 kratos2 bluetoothd[81989]: src/adapter.c:connected_callback() hci0 device F2:1E:0F:33:13:1D connected eir_len 25
Jan  4 22:52:18 kratos2 bluetoothd[81989]: src/device.c:att_connect_cb() connect to F2:1E:0F:33:13:1D: Software caused connection abort (103)
Jan  4 22:52:18 kratos2 bluetoothd[81989]: src/adapter.c:dev_disconnected() Device F2:1E:0F:33:13:1D disconnected, reason 4
Jan  4 22:52:18 kratos2 bluetoothd[81989]: src/adapter.c:adapter_remove_connection() 
Jan  4 22:52:18 kratos2 bluetoothd[81989]: plugins/policy.c:disconnect_cb() reason 4
Jan  4 22:52:18 kratos2 bluetoothd[81989]: src/adapter.c:bonding_attempt_complete() hci0 bdaddr F2:1E:0F:33:13:1D type 2 status 0xe
Jan  4 22:52:18 kratos2 bluetoothd[81989]: src/device.c:device_bonding_complete() bonding (nil) status 0x0e
Jan  4 22:52:18 kratos2 bluetoothd[81989]: src/device.c:device_bonding_failed() status 14
Jan  4 22:52:18 kratos2 bluetoothd[81989]: src/adapter.c:resume_discovery() 
Jan  4 22:52:20 kratos2 bluetoothd[81989]: src/adapter.c:connected_callback() hci0 device F2:1E:0F:33:13:1D connected eir_len 25
Jan  4 22:52:20 kratos2 bluetoothd[81989]: src/device.c:browse_request_cancel() 
Jan  4 22:52:20 kratos2 bluetoothd[81989]: src/gatt-database.c:connect_cb() New incoming LE ATT connection
Jan  4 22:52:20 kratos2 bluetoothd[81989]: src/device.c:device_attach_att() Elevating security level since LTK is available
Jan  4 22:52:20 kratos2 bluetoothd[81989]: attrib/gattrib.c:g_attrib_ref() 0x55feefa00cb0: g_attrib_ref=1 
Jan  4 22:52:20 kratos2 bluetoothd[81989]: profiles/gap/gas.c:gap_accept() GAP profile accept (F2:1E:0F:33:13:1D)
Jan  4 22:52:20 kratos2 bluetoothd[81989]: profiles/gap/gas.c:handle_characteristic() Unsupported characteristic: 00002a04-0000-1000-8000-00805f9b34fb
Jan  4 22:52:20 kratos2 bluetoothd[81989]: src/service.c:change_state() 0x55feefa04310: device F2:1E:0F:33:13:1D profile gap-profile state changed: disconnected -> connected (0)
Jan  4 22:52:20 kratos2 bluetoothd[81989]: src/gatt-client.c:btd_gatt_client_connected() Device connected.
Jan  4 22:52:20 kratos2 bluetoothd[81989]: src/adapter.c:dev_disconnected() Device F2:1E:0F:33:13:1D disconnected, reason 4
Jan  4 22:52:20 kratos2 bluetoothd[81989]: src/adapter.c:adapter_remove_connection() 
Jan  4 22:52:20 kratos2 bluetoothd[81989]: plugins/policy.c:disconnect_cb() reason 4
Jan  4 22:52:20 kratos2 bluetoothd[81989]: src/adapter.c:bonding_attempt_complete() hci0 bdaddr F2:1E:0F:33:13:1D type 2 status 0xe
Jan  4 22:52:20 kratos2 bluetoothd[81989]: src/device.c:device_bonding_complete() bonding (nil) status 0x0e
Jan  4 22:52:20 kratos2 bluetoothd[81989]: src/device.c:device_bonding_failed() status 14
Jan  4 22:52:20 kratos2 bluetoothd[81989]: src/adapter.c:resume_discovery() 
Jan  4 22:52:20 kratos2 bluetoothd[81989]: src/device.c:gatt_debug() Channel 0x55feefa172a0 disconnected: Software caused connection abort
Jan  4 22:52:20 kratos2 bluetoothd[81989]: src/device.c:gatt_debug() MTU Exchange failed. ATT ECODE: 0x00
Jan  4 22:52:20 kratos2 bluetoothd[81989]: src/device.c:gatt_client_ready_cb() status: failed, error: 0
Jan  4 22:52:20 kratos2 bluetoothd[81989]: src/device.c:device_svc_resolved() /org/bluez/hci0/dev_F2_1E_0F_33_13_1D err -5
Jan  4 22:52:20 kratos2 bluetoothd[81989]: profiles/gap/gas.c:read_device_name_cb() Reading device name failed with ATT errror: 14
Jan  4 22:52:20 kratos2 bluetoothd[81989]: profiles/gap/gas.c:read_appearance_cb() Reading appearance failed with ATT error: 14
Jan  4 22:52:20 kratos2 bluetoothd[81989]: src/device.c:att_disconnected_cb() 
Jan  4 22:52:20 kratos2 bluetoothd[81989]: src/device.c:att_disconnected_cb() Software caused connection abort (103)
Jan  4 22:52:20 kratos2 bluetoothd[81989]: src/service.c:change_state() 0x55feefa04310: device F2:1E:0F:33:13:1D profile gap-profile state changed: connected -> disconnecting (0)
Jan  4 22:52:20 kratos2 bluetoothd[81989]: src/service.c:change_state() 0x55feefa04310: device F2:1E:0F:33:13:1D profile gap-profile state changed: disconnecting -> disconnected (0)
Jan  4 22:52:20 kratos2 bluetoothd[81989]: src/gatt-client.c:btd_gatt_client_disconnected() Device disconnected. Cleaning up.
Jan  4 22:52:20 kratos2 bluetoothd[81989]: src/device.c:att_disconnected_cb() Automatic connection disabled
Jan  4 22:52:20 kratos2 bluetoothd[81989]: src/gatt-database.c:btd_gatt_database_att_disconnected() 
Jan  4 22:52:20 kratos2 bluetoothd[81989]: attrib/gattrib.c:g_attrib_unref() 0x55feefa00cb0: g_attrib_unref=0 
Jan  4 22:52:21 kratos2 bluetoothd[81989]: src/device.c:device_connect_le() Connection attempt to: F2:1E:0F:33:13:1D
Jan  4 22:52:22 kratos2 bluetoothd[81989]: src/adapter.c:connected_callback() hci0 device F2:1E:0F:33:13:1D connected eir_len 25
Jan  4 22:52:23 kratos2 bluetoothd[81989]: src/device.c:att_connect_cb() connect to F2:1E:0F:33:13:1D: Software caused connection abort (103)
Jan  4 22:52:23 kratos2 bluetoothd[81989]: src/adapter.c:dev_disconnected() Device F2:1E:0F:33:13:1D disconnected, reason 4
Jan  4 22:52:23 kratos2 bluetoothd[81989]: src/adapter.c:adapter_remove_connection() 
Jan  4 22:52:23 kratos2 bluetoothd[81989]: plugins/policy.c:disconnect_cb() reason 4
Jan  4 22:52:23 kratos2 bluetoothd[81989]: src/adapter.c:bonding_attempt_complete() hci0 bdaddr F2:1E:0F:33:13:1D type 2 status 0xe
Jan  4 22:52:23 kratos2 bluetoothd[81989]: src/device.c:device_bonding_complete() bonding (nil) status 0x0e
Jan  4 22:52:23 kratos2 bluetoothd[81989]: src/device.c:device_bonding_failed() status 14
Jan  4 22:52:23 kratos2 bluetoothd[81989]: src/adapter.c:resume_discovery() 
Jan  4 22:52:25 kratos2 bluetoothd[81989]: src/adapter.c:connected_callback() hci0 device F2:1E:0F:33:13:1D connected eir_len 25
Jan  4 22:52:25 kratos2 bluetoothd[81989]: src/device.c:device_attach_att() Elevating security level since LTK is available
Jan  4 22:52:25 kratos2 bluetoothd[81989]: attrib/gattrib.c:g_attrib_ref() 0x55feefa09bd0: g_attrib_ref=1 
Jan  4 22:52:25 kratos2 bluetoothd[81989]: profiles/gap/gas.c:gap_accept() GAP profile accept (F2:1E:0F:33:13:1D)
Jan  4 22:52:25 kratos2 bluetoothd[81989]: profiles/gap/gas.c:handle_characteristic() Unsupported characteristic: 00002a04-0000-1000-8000-00805f9b34fb
Jan  4 22:52:25 kratos2 bluetoothd[81989]: src/service.c:change_state() 0x55feefa04310: device F2:1E:0F:33:13:1D profile gap-profile state changed: disconnected -> connected (0)
Jan  4 22:52:25 kratos2 bluetoothd[81989]: src/gatt-client.c:btd_gatt_client_connected() Device connected.
Jan  4 22:52:25 kratos2 bluetoothd[81989]: src/adapter.c:dev_disconnected() Device F2:1E:0F:33:13:1D disconnected, reason 4
Jan  4 22:52:25 kratos2 bluetoothd[81989]: src/adapter.c:adapter_remove_connection() 
Jan  4 22:52:25 kratos2 bluetoothd[81989]: plugins/policy.c:disconnect_cb() reason 4
Jan  4 22:52:25 kratos2 bluetoothd[81989]: src/adapter.c:bonding_attempt_complete() hci0 bdaddr F2:1E:0F:33:13:1D type 2 status 0xe
Jan  4 22:52:25 kratos2 bluetoothd[81989]: src/device.c:device_bonding_complete() bonding (nil) status 0x0e
Jan  4 22:52:25 kratos2 bluetoothd[81989]: src/device.c:device_bonding_failed() status 14
Jan  4 22:52:25 kratos2 bluetoothd[81989]: src/adapter.c:resume_discovery() 
Jan  4 22:52:25 kratos2 bluetoothd[81989]: src/device.c:gatt_debug() Channel 0x55feefa172a0 disconnected: Software caused connection abort
Jan  4 22:52:25 kratos2 bluetoothd[81989]: src/device.c:gatt_debug() MTU Exchange failed. ATT ECODE: 0x00
Jan  4 22:52:25 kratos2 bluetoothd[81989]: src/device.c:gatt_client_ready_cb() status: failed, error: 0
Jan  4 22:52:25 kratos2 bluetoothd[81989]: src/device.c:device_svc_resolved() /org/bluez/hci0/dev_F2_1E_0F_33_13_1D err -5
Jan  4 22:52:25 kratos2 bluetoothd[81989]: profiles/gap/gas.c:read_device_name_cb() Reading device name failed with ATT errror: 14
Jan  4 22:52:25 kratos2 bluetoothd[81989]: profiles/gap/gas.c:read_appearance_cb() Reading appearance failed with ATT error: 14
Jan  4 22:52:25 kratos2 bluetoothd[81989]: src/device.c:att_disconnected_cb() 
Jan  4 22:52:25 kratos2 bluetoothd[81989]: src/device.c:att_disconnected_cb() Software caused connection abort (103)
Jan  4 22:52:25 kratos2 bluetoothd[81989]: src/service.c:change_state() 0x55feefa04310: device F2:1E:0F:33:13:1D profile gap-profile state changed: connected -> disconnecting (0)
Jan  4 22:52:25 kratos2 bluetoothd[81989]: src/service.c:change_state() 0x55feefa04310: device F2:1E:0F:33:13:1D profile gap-profile state changed: disconnecting -> disconnected (0)
Jan  4 22:52:25 kratos2 bluetoothd[81989]: src/gatt-client.c:btd_gatt_client_disconnected() Device disconnected. Cleaning up.
Jan  4 22:52:25 kratos2 bluetoothd[81989]: src/device.c:att_disconnected_cb() Automatic connection disabled
Jan  4 22:52:25 kratos2 bluetoothd[81989]: src/gatt-database.c:btd_gatt_database_att_disconnected() 
Jan  4 22:52:25 kratos2 bluetoothd[81989]: attrib/gattrib.c:g_attrib_unref() 0x55feefa09bd0: g_attrib_unref=0 
Jan  4 22:52:34 kratos2 bluetoothd[81989]: src/device.c:device_connect_le() Connection attempt to: F2:1E:0F:33:13:1D
Jan  4 22:52:34 kratos2 bluetoothd[81989]: src/adapter.c:connected_callback() hci0 device F2:1E:0F:33:13:1D connected eir_len 25
Jan  4 22:52:35 kratos2 bluetoothd[81989]: src/device.c:att_connect_cb() connect to F2:1E:0F:33:13:1D: Software caused connection abort (103)
Jan  4 22:52:35 kratos2 bluetoothd[81989]: src/adapter.c:dev_disconnected() Device F2:1E:0F:33:13:1D disconnected, reason 4
Jan  4 22:52:35 kratos2 bluetoothd[81989]: src/adapter.c:adapter_remove_connection() 
Jan  4 22:52:35 kratos2 bluetoothd[81989]: plugins/policy.c:disconnect_cb() reason 4
Jan  4 22:52:35 kratos2 bluetoothd[81989]: src/adapter.c:bonding_attempt_complete() hci0 bdaddr F2:1E:0F:33:13:1D type 2 status 0xe
Jan  4 22:52:35 kratos2 bluetoothd[81989]: src/device.c:device_bonding_complete() bonding (nil) status 0x0e
Jan  4 22:52:35 kratos2 bluetoothd[81989]: src/device.c:device_bonding_failed() status 14
Jan  4 22:52:35 kratos2 bluetoothd[81989]: src/adapter.c:resume_discovery() 
Jan  4 22:52:37 kratos2 bluetoothd[81989]: src/adapter.c:connected_callback() hci0 device F2:1E:0F:33:13:1D connected eir_len 25
Jan  4 22:52:37 kratos2 bluetoothd[81989]: src/device.c:device_attach_att() Elevating security level since LTK is available
Jan  4 22:52:37 kratos2 bluetoothd[81989]: attrib/gattrib.c:g_attrib_ref() 0x55feef9fef70: g_attrib_ref=1 
Jan  4 22:52:37 kratos2 bluetoothd[81989]: profiles/gap/gas.c:gap_accept() GAP profile accept (F2:1E:0F:33:13:1D)
Jan  4 22:52:37 kratos2 bluetoothd[81989]: profiles/gap/gas.c:handle_characteristic() Unsupported characteristic: 00002a04-0000-1000-8000-00805f9b34fb
Jan  4 22:52:37 kratos2 bluetoothd[81989]: src/service.c:change_state() 0x55feefa04310: device F2:1E:0F:33:13:1D profile gap-profile state changed: disconnected -> connected (0)
Jan  4 22:52:37 kratos2 bluetoothd[81989]: src/gatt-client.c:btd_gatt_client_connected() Device connected.
Jan  4 22:52:37 kratos2 bluetoothd[81989]: src/adapter.c:dev_disconnected() Device F2:1E:0F:33:13:1D disconnected, reason 3
Jan  4 22:52:37 kratos2 bluetoothd[81989]: src/adapter.c:adapter_remove_connection() 
Jan  4 22:52:37 kratos2 bluetoothd[81989]: plugins/policy.c:disconnect_cb() reason 3
Jan  4 22:52:37 kratos2 bluetoothd[81989]: src/adapter.c:bonding_attempt_complete() hci0 bdaddr F2:1E:0F:33:13:1D type 2 status 0xe
Jan  4 22:52:37 kratos2 bluetoothd[81989]: src/device.c:device_bonding_complete() bonding (nil) status 0x0e
Jan  4 22:52:37 kratos2 bluetoothd[81989]: src/device.c:device_bonding_failed() status 14
Jan  4 22:52:37 kratos2 bluetoothd[81989]: src/adapter.c:resume_discovery() 
Jan  4 22:52:37 kratos2 bluetoothd[81989]: src/device.c:gatt_debug() Channel 0x55feefa172a0 disconnected: Connection reset by peer
Jan  4 22:52:37 kratos2 bluetoothd[81989]: src/device.c:gatt_debug() MTU Exchange failed. ATT ECODE: 0x00
Jan  4 22:52:37 kratos2 bluetoothd[81989]: src/device.c:gatt_client_ready_cb() status: failed, error: 0
Jan  4 22:52:37 kratos2 bluetoothd[81989]: src/device.c:device_svc_resolved() /org/bluez/hci0/dev_F2_1E_0F_33_13_1D err -5
Jan  4 22:52:37 kratos2 bluetoothd[81989]: profiles/gap/gas.c:read_device_name_cb() Reading device name failed with ATT errror: 14
Jan  4 22:52:37 kratos2 bluetoothd[81989]: profiles/gap/gas.c:read_appearance_cb() Reading appearance failed with ATT error: 14
Jan  4 22:52:37 kratos2 bluetoothd[81989]: src/device.c:att_disconnected_cb() 
Jan  4 22:52:37 kratos2 bluetoothd[81989]: src/device.c:att_disconnected_cb() Connection reset by peer (104)
Jan  4 22:52:37 kratos2 bluetoothd[81989]: src/service.c:change_state() 0x55feefa04310: device F2:1E:0F:33:13:1D profile gap-profile state changed: connected -> disconnecting (0)
Jan  4 22:52:37 kratos2 bluetoothd[81989]: src/service.c:change_state() 0x55feefa04310: device F2:1E:0F:33:13:1D profile gap-profile state changed: disconnecting -> disconnected (0)
Jan  4 22:52:37 kratos2 bluetoothd[81989]: src/gatt-client.c:btd_gatt_client_disconnected() Device disconnected. Cleaning up.
Jan  4 22:52:37 kratos2 bluetoothd[81989]: src/device.c:att_disconnected_cb() Automatic connection disabled
Jan  4 22:52:37 kratos2 bluetoothd[81989]: src/gatt-database.c:btd_gatt_database_att_disconnected() 
Jan  4 22:52:37 kratos2 bluetoothd[81989]: attrib/gattrib.c:g_attrib_unref() 0x55feef9fef70: g_attrib_unref=0 
```

**Comment by cagnulein** [comment]
hi @loikki not all the bluetooth dongle are able to use bluetooth central and peripheral at the same time.
try to disable the virtual device settings in the experimental settings and restart qz 

if the problem with the bike disappears, i'm right :) 

let me know

**Comment by loikki** [comment]
Thanks for your help. I managed to connect. I disabled the virtual device as you said but it was not sufficient. I had to unpair and forget my bike in the bluetooth settings before launching your software.

By the way, I am planning to write my own script that generates a training based on the amount of energy I want to spend (and a few other parameters). If you are interested, I will be happy to share them with you once they are ready.

Feel free to close the issue.
